### PR TITLE
Add configuration validation

### PR DESCRIPTION
**Pull request by AaronDavidSchneider**
One step towards ha-fritzbox-tools as a standard component.
`device_list` is replaced by `devices`, as supplied by homeassistant.const.

Works with my setup.

One question remains for me:
Is username an optional parameter? Do the apis work with no username?

Edit - Todo towards correct styles:
- [x] change formating strings
```
# New
f"{some_value} {some_other_value}"
# Old, wrong
"{} {}".format('New', 'style')
"%s %s" % ('Old', 'style')
```
- [x] replace header by meaningful names
- [x] decide on username as optional/required

All requirements:
https://developers.home-assistant.io/docs/en/development_guidelines.html
https://developers.home-assistant.io/docs/en/creating_component_code_review.html

### CODE REVIEW

File: custom_components/fritzbox_tools/switch.py [hunk 1]
```diff
@@ -85,12 +87,8 @@ def __init__(self, fritzbox_tools, port_mapping, idx, connection_type):
         self.connection_type = connection_type
         self.port_mapping: dict = port_mapping  # dict in the format as it comes from fritzconnection. eg: {'NewRemoteHost': '0.0.0.0', 'NewExternalPort': 22, 'NewProtocol': 'TCP', 'NewInternalPort': 22, 'NewInternalClient': '192.168.178.31', 'NewEnabled': '0', 'NewPortMappingDescription': 'Beast SSH ', 'NewLeaseDuration': 0}  # noqa
 
-        self._name = "Port forward {}".format(
-            port_mapping["NewPortMappingDescription"]
-        )
-        id = "fritzbox_portforward_{}".format(
-            slugify(port_mapping["NewPortMappingDescription"])
-        )
+        self._name = f'Port forward {port_mapping['NewPortMappingDescription']}'
```
Comment (mammuth) [inline custom_components/fritzbox_tools/switch.py]: Ah, I first thought we could not use f-strings yet, but it turns out, Home Assistant doesn't support Python 3.5 anymore, so we're good to go!

File: custom_components/fritzbox_tools/__init__.py [hunk 2]
```diff
@@ -1,32 +1,76 @@
...
+
 DOMAIN = 'fritzbox_tools'
 SUPPORTED_DOMAINS = ['switch', 'sensor']
 
+CONF_PROFILE_ON = 'profile_on'
+CONF_PROFILE_OFF = 'profile_off'
+CONF_HOMEASSISTANT_IP = 'homeassistant_ip'
+
+DEFAULT_PROFILE_OFF = 'Gesperrt'
+DEFAULT_HOST = '192.168.178.1' #set to fritzbox default
```
Comment (mammuth) [inline custom_components/fritzbox_tools/__init__.py]: The CI should complain about the missing gaps, so obviously it's configured wrongly 😏

File: custom_components/fritzbox_tools/switch.py (hunk 1 above)
Comment (mammuth) [reply custom_components/fritzbox_tools/switch.py]: Syntax error due to nested single quotes. I'd use double quotes for the inner usage.

File: custom_components/fritzbox_tools/switch.py [hunk 3]
```diff
@@ -86,12 +87,9 @@ def __init__(self, fritzbox_tools, port_mapping, idx, connection_type):
...
 
-        self._name = "Port forward {}".format(
-            port_mapping["NewPortMappingDescription"]
-        )
-        id = "fritzbox_portforward_{}".format(
-            slugify(port_mapping["NewPortMappingDescription"])
-        )
+        description = port_mapping['NewPortMappingDescription']
+        self._name = f'Port forward {description}'
+        id = f'fritzbox_portforward_{slugify(description}'
```
Comment (mammuth) [inline custom_components/fritzbox_tools/switch.py]: Missing `)`

**Comment by mammuth** [comment]
Cool. According to a manual test in my setup it works.

Would you mind fixing the open lint errors before we merge it? 
```
$ git diff master...config_validation | flake8 --diff .  
```
```python
custom_components/fritzbox_tools/__init__.py:27:31: E261 at least two spaces before inline comment
custom_components/fritzbox_tools/__init__.py:27:32: E262 inline comment should start with '# '
custom_components/fritzbox_tools/__init__.py:28:21: E261 at least two spaces before inline comment
custom_components/fritzbox_tools/__init__.py:28:22: E262 inline comment should start with '# '
custom_components/fritzbox_tools/__init__.py:29:22: E261 at least two spaces before inline comment
custom_components/fritzbox_tools/__init__.py:29:23: E262 inline comment should start with '# '
custom_components/fritzbox_tools/__init__.py:51:56: E261 at least two spaces before inline comment
custom_components/fritzbox_tools/__init__.py:91:5: E303 too many blank lines (2)
custom_components/fritzbox_tools/__init__.py:113:1: W293 blank line contains whitespace
```

**Comment by AaronDavidSchneider** [comment]
I get shown different ones. I will resolve them later.

**Comment by mammuth** [comment]
I removed unrelated issues in the README.md (which is excluded in the master)

**Comment by AaronDavidSchneider** [comment]
have you decided on the headers yet?

**Comment by mammuth** [comment]
> please also comment on username as requirement or not

Apparently on my setup some things work when using the "root password" instead of a username / password combination.

However, for example the guest WiFi switch only works when the component is configured with a user (it's being marked as unavailable).
Also, the following authorization error occurs for the port switches when running with password-only:
```
2019-10-29 13:56:57 ERROR (MainThread) [custom_components.fritzbox_tools.switch] Port switches could not be enabled. Check if your fritzbox is able to do port forwardings!
Traceback (most recent call last):
  File "/Users/mammuth/.homeassistant/custom_components/fritzbox_tools/switch.py", line 30, in async_setup_platform
    'GetDefaultConnectionService'
  File "/Users/mammuth/.homeassistant/deps/lib/python/site-packages/fritzconnection/fritzconnection.py", line 490, in call_action
    return action.execute(**kwargs)
  File "/Users/mammuth/.homeassistant/deps/lib/python/site-packages/fritzconnection/fritzconnection.py", line 171, in execute
    raise AuthorizationError('unauthorized request')
```
The switches seem to be enabled anyways and I can also change their state 🤔

(For reference, I've got the setting `Anmeldung mit dem FRITZ!Box-Kennwort` enabled (`System > FRITZ!Box-Benutzer > Anmeldung im Heimnetz`))

So, maybe let's stick with having the username required, to avoid issues due to the authentication?

> have you decided on the headers yet?
I'm fine with the file headers you added 👍

**Comment by mammuth** [comment]
Alright, I manually tested the changes on my setup (we should really consider creating automated tests...).

I'll merge it if you say it's ready to go?

**Pull request merged**